File: src/main/java/com/fwdekker/randomness/SettingsManager.java
package com.fwdekker.randomness;

import org.jetbrains.annotations.NotNull;


/**
 * A {@code SettingsManager} is an object that can change settings of a given type.
 *
 * @param <S> the type of settings that are changed by the implementation
 */
public interface SettingsManager<S extends Settings> {
    /**
     * Loads the default settings instance into the implementing object.
     */
    void loadSettings();

    /**
     * Loads the given settings instance into the implementing object.
     *
     * @param settings a settings instance
     */
    void loadSettings(@NotNull S settings);

    /**
     * Saves the implementing object's state into the default settings instance.
     */
    void saveSettings();

    /**
     * Saves the implementing object's state into the given settings instance.
     *
     * @param settings a settings instance
     */
    void saveSettings(@NotNull S settings);
}

File: src/main/kotlin/com/fwdekker/randomness/SettingsManager.kt
package com.fwdekker.randomness


/**
 * A `SettingsManager` is an object that can save and load settings.
 *
 * @param S the type of settings that is saved and loaded
 * @param T the type of scheme that is contained within the settings
 */
interface SettingsManager<S : Settings<*, T>, T : Scheme<T>> {
    /**
     * Loads the default settings object.
     */
    fun loadSettings()

    /**
     * Loads `settings`.
     *
     * @param settings the settings to load
     */
    fun loadSettings(settings: S)

    /**
     * Saves settings to the default settings object.
     */
    fun saveSettings()

    /**
     * Saves settings to `settings`.
     *
     * @param settings the settings to save to
     */
    fun saveSettings(settings: S)
}
